File: platforms/iPad/Accounter iOS 1.0.2/Accounter/ResultJSONDeserializer.m
//
//  ResultJSONDeserializer.m
//  Accounter iOS2
//
//  Created by Amrit Mishra on 11/11/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "ResultJSONDeserializer.h"


@implementation ResultJSONDeserializer



-(id)init{
    
    self = [super init];
    if (self) {
        
        
    }
    return self;
}

-(void)setObject :(id)newJSONObject{
    
    
    jsonObject = newJSONObject;

}


-(Result*)newDeserializer{
    
    Result *result = [[Result alloc]init];
    
    NSDictionary *jsonObjectDict = [[NSDictionary alloc]init];
    jsonObjectDict = jsonObject;
    
    if([jsonObjectDict valueForKey:@"cookie"]!=nil)
    {
        NSString *cookieString = [[NSString alloc]initWithString:[jsonObject valueForKey:@"cookie"]];
        NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
        [prefs setObject:cookieString forKey:@"accounterCookie"];
        
        [cookieString release];
        
    }
    
    BOOL value = [[jsonObjectDict valueForKey:@"hideCancel"]boolValue];
    if(value == false)
    {
        [result setHideCancel:false];
    }else{
        [result setHideCancel:true];
    }
    
    BOOL value1 = [[jsonObjectDict valueForKey:@"showBack"]boolValue];
    if(value1 == false)
    {
        [result setShowBack:true];
    }else{
        [result setShowBack:false];
    }
    
    
    if([jsonObjectDict valueForKey:@"title"]!=nil)
    {
        [result setTitle:[jsonObjectDict valueForKey:@"title"]];
    }
    
    
    NSMutableArray *jresultParts = [jsonObjectDict valueForKey:@"resultParts"];
    NSMutableArray * resultParts = [[NSMutableArray alloc]init];
    resultParts = [result getResultParts];
    
    for(int i = 0;i<[jresultParts count];i++){
        id asJsonObject = [jresultParts objectAtIndex:i];
        
        int type = [[asJsonObject valueForKey:@"type"]intValue];
        
        switch (type) {
            case 0:{
                NSString *asString = [[NSString alloc]initWithString:[asJsonObject valueForKey:@"message"]];
                [resultParts addObject:asString];
            }break;
                
            case 1:{
                ResultList *list = [[ResultList alloc]init];
                
                BOOL value = [[asJsonObject valueForKey:@"isMultiSelection"]boolValue];
                [list setMutiSelection:value];
                
                [list setTitle:[asJsonObject valueForKey:@"title"]];
                
                if ([asJsonObject valueForKey:@"isMultiSelection"] == 0) {
                    [list setMutiSelection:FALSE];
                }else{
                    [list setMutiSelection:TRUE];
                }
                [list setName:[asJsonObject valueForKey:@"name"]];
                
                NSMutableArray *records = [[NSMutableArray alloc]init ];
                records =[list getRecords];
                NSMutableArray *asJsonArray = [asJsonObject valueForKey:@"records"];
                
                for (int k =0; k <[asJsonArray count]; k++) {
                    
                    NSDictionary *asJsonObject2 = [[NSDictionary alloc]init];
                    asJsonObject2 = [asJsonArray objectAtIndex:k];
                    
                    
                    Record *record = [[Record alloc]init];
                    [record setCode:[asJsonObject2 valueForKey:@"code"]];
                    
                    NSArray *asJsonArray2 = [asJsonObject2 valueForKey:@"cells"];
                    NSMutableArray *cells = [[NSMutableArray alloc]init];
                    
                    for (int z =0; z < [asJsonArray2 count]; z++) {
                        
                        NSDictionary *asJsonObject3 = [[NSDictionary alloc]init];
                        asJsonObject3  = [asJsonArray2 objectAtIndex:z];
                        
                        Cell *cell = [[Cell alloc]init];
                        [cell setName:[asJsonObject3 valueForKey:@"title"]];
                        [cell setValue:[asJsonObject3 valueForKey:@"value"]];
                        
                        [cells addObject:cell];
                        
                    }
                    [record setCells:cells];
                    [records addObject:record];
                }
                [list setRecords:records];
                [resultParts addObject:list];
                
            }break;
                
            case 2:{
                
                NSMutableArray *asJsonArray = [asJsonObject valueForKey:@"commandNames"];
                if(asJsonArray != nil){
                    
                    CommandList *commandsList = [[CommandList alloc]init];
                    
                    NSMutableArray *commandNames = [[NSMutableArray alloc]initWithArray:[commandsList getCommandNames]];
                    
                    for (int j =0; j < [asJsonArray count];j++) {
                        
                        NSDictionary *asJsonObject2 = [asJsonArray objectAtIndex:j];
                        
                        Command *command = [[Command alloc]init];
                        [command setName:[asJsonObject2 valueForKey:@"name"]];
                        [command setCode:[asJsonObject2 valueForKey:@"code"]];
                        
                        [commandNames addObject:command];
                    }
                    [commandsList setCommadNames:commandNames];
                    [resultParts addObject:commandsList];
                    // [commandNames release];
                }
                
            }break;
                
            case 3:{
                int input =  [[asJsonObject valueForKey:@"inputType"]intValue];
                
                InputType *inputType = [[InputType alloc]init];
                [inputType setInputType:input];
                
                NSString *name = [[NSString alloc]initWithString:[asJsonObject valueForKey:@"name"]];
                [inputType setName:name];
                
                NSString *value = [[NSString alloc]initWithString:[asJsonObject valueForKey:@"value"]];
                [inputType setTextValue:value];
                
                [resultParts addObject:inputType];
                
            }break;
            default:
                break;
        }
        
    }
    
    
    return  result;
}


/*
 -(Result *)getDeserializedObject{
 
 Result *result = [[Result alloc]init];
 
 NSDictionary *jsonObjectDict = [[NSDictionary alloc]init];
 jsonObjectDict = jsonObject;
 
 
 if([jsonObjectDict valueForKey:@"cookie"]!=nil)
 {
 NSString *cookieString = [[NSString alloc]initWithString:[jsonObject valueForKey:@"cookie"]];
 NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
 [prefs setObject:cookieString forKey:@"accounterCookie"];
 
 }
 
 
 NSMutableArray *jresultParts = [jsonObjectDict valueForKey:@"resultParts"];
 
 NSMutableArray * resultParts = [[NSMutableArray alloc]init];
 resultParts = [result getResultParts];
 
 
 for(int i = 0;i<[jresultParts count];i++){
 
 id jsonElement = [jresultParts objectAtIndex:i];
 
 if ([[jsonElement class] isSubclassOfClass:[NSString class]]) {
 NSString *asString = [[NSString alloc]init];
 asString = jsonElement;
 [resultParts addObject:asString];
 }else{
 NSDictionary *asJsonObject = [[NSDictionary alloc]init];
 asJsonObject  = jsonElement;
 
 NSMutableArray *asJsonArray = [asJsonObject valueForKey:@"commandNames"];
 if(asJsonArray != nil){
 
 CommandList *commandsList = [[CommandList alloc]init];
 
 NSMutableArray *commandNames = [[NSMutableArray alloc]initWithArray:[commandsList getCommandNames]];
 
 for (int j =0; j < [asJsonArray count];j++) {
 
 NSDictionary *asJsonObject2 = [asJsonArray objectAtIndex:j];
 
 Command *command = [[Command alloc]init];
 [command setName:[asJsonObject2 valueForKey:@"name"]];
 [command setCode:[asJsonObject2 valueForKey:@"code"]];
 
 [commandNames addObject:command];
 }
 [commandsList setCommadNames:commandNames];
 [resultParts addObject:commandsList];
 
 }else{
 
 ResultList *list = [[ResultList alloc]init];
 
 if ([asJsonObject valueForKey:@"isMultiSelection"] == 0) {
 [list setMutiSelection:FALSE];
 }else{
 [list setMutiSelection:TRUE];
 }
 [list setName:[asJsonObject valueForKey:@"name"]];
 
 NSMutableArray *records = [[NSMutableArray alloc]init ];
 records =[list getRecords];
 NSMutableArray *asJsonArray = [asJsonObject valueForKey:@"records"];
 
 for (int k =0; k <[asJsonArray count]; k++) {
 
 NSDictionary *asJsonObject2 = [[NSDictionary alloc]init];
 asJsonObject2 = [asJsonArray objectAtIndex:k];
 
 
 Record *record = [[Record alloc]init];
 [record setCode:[asJsonObject2 valueForKey:@"code"]];
 
 NSArray *asJsonArray2 = [asJsonObject2 valueForKey:@"cells"];
 NSMutableArray *cells = [[NSMutableArray alloc]init];
 
 for (int z =0; z < [asJsonArray2 count]; z++) {
 
 NSDictionary *asJsonObject3 = [[NSDictionary alloc]init];
 asJsonObject3  = [asJsonArray2 objectAtIndex:z];
 
 Cell *cell = [[Cell alloc]init];
 [cell setName:[asJsonObject3 valueForKey:@"name"]];
 [cell setValue:[asJsonObject3 valueForKey:@"value"]];
 
 [cells addObject:cell];
 
 }
 [record setCells:cells];
 [records addObject:record];
 }
 [list setRecords:records];
 [resultParts addObject:list];
 }
 }
 
 
 }
 if([jsonObjectDict valueForKey:@"inputType"]!=nil)
 {
 NSString *value = [jsonObjectDict valueForKey:@"inputType"];
 [result setInputType:[value intValue]];        
 }
 
 if([jsonObjectDict valueForKey:@"hideCancel"]!=nil)
 {
 
 BOOL value = [[jsonObjectDict valueForKey:@"hideCancel"]boolValue];
 if(value == false)
 {
 [result setHideCancel:false];
 }else{
 [result setHideCancel:true];
 }
 }
 
 if([jsonObjectDict valueForKey:@"showBack"]!=nil)
 {
 BOOL value = [[jsonObjectDict valueForKey:@"showBack"]boolValue];
 if(value == false)
 {
 [result setShowBack:true];
 }else{
 [result setShowBack:false];
 }
 }
 
 if([jsonObjectDict valueForKey:@"title"]!=nil)
 {
 [result setTitle:[jsonObjectDict valueForKey:@"title"]];
 }
 [result setResultParts:resultParts];
 return  result;
 }
 */
@end

File: platforms/Android/Accounter/src/com/vimukti/accounter/result/ResultJsonDeserializer.java
package com.vimukti.accounter.result;

import java.lang.reflect.Type;
import java.util.List;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

public class ResultJsonDeserializer implements JsonDeserializer<Result> {

	@Override
	public Result deserialize(JsonElement json, Type arg1,
			JsonDeserializationContext arg2) throws JsonParseException {
		Result result = new Result();
		JsonObject jsonObject = json.getAsJsonObject();
		JsonElement jsonElement3 = jsonObject.get("cookie");
		if (jsonElement3 != null) {
			result.setCookie(jsonElement3.getAsString());
		}
		JsonElement title = jsonObject.get("title");
		if (title != null) {
			result.setTitle(title.getAsString());
		}

		result.setHideCancel(jsonObject.get("hideCancel").getAsBoolean());
		result.setShowBack(jsonObject.get("showBack").getAsBoolean());

		JsonArray jresultParts = jsonObject.get("resultParts").getAsJsonArray();
		List<Object> resultParts = result.getResultParts();
		for (int i = 0; i < jresultParts.size(); i++) {
			JsonObject asJsonObject = jresultParts.get(i).getAsJsonObject();
			int type = asJsonObject.get("type").getAsInt();
			switch (type) {
			case 0:
				String asString = asJsonObject.get("message").getAsString();
				resultParts.add(asString);
				break;
			case 1:
				ResultList list = new ResultList();
				list.setMultiSelection(asJsonObject.get("isMultiSelection")
						.getAsBoolean());
				JsonElement resultTitle = asJsonObject.get("title");
				if (resultTitle != null) {
					list.setTitle(resultTitle.getAsString());
				}
				JsonArray asJsonArray = asJsonObject.get("records")
						.getAsJsonArray();
				List<Record> records = list.getRecords();
				for (int j = 0; j < asJsonArray.size(); j++) {
					JsonObject asJsonObject2 = asJsonArray.get(j)
							.getAsJsonObject();
					Record record = new Record();
					String code = asJsonObject2.get("code").getAsString();
					record.setCode(code);
					JsonArray asJsonArray2 = asJsonObject2.get("cells")
							.getAsJsonArray();
					List<Cell> cells = record.getCells();
					for (int k = 0; k < asJsonArray2.size(); k++) {
						JsonObject asJsonObject3 = asJsonArray2.get(k)
								.getAsJsonObject();
						Cell cell = new Cell();
						cell.setTitle(asJsonObject3.get("title").getAsString());
						cell.setValue(asJsonObject3.get("value").getAsString());
						cells.add(cell);
					}
					records.add(record);
				}
				resultParts.add(list);
				break;
			case 2:
				JsonElement jsonElement2 = asJsonObject.get("commandNames");
				CommandList commandList = new CommandList();
				List<Command> commandNames = commandList.getCommandNames();
				JsonArray asJsonArray2 = jsonElement2.getAsJsonArray();
				for (int j = 0; j < asJsonArray2.size(); j++) {
					JsonObject asJsonObject2 = asJsonArray2.get(j)
							.getAsJsonObject();
					Command command = new Command();
					command.setName(asJsonObject2.get("name").getAsString());
					command.setCode(asJsonObject2.get("code").getAsString());
					commandNames.add(command);
				}
				resultParts.add(commandList);
				break;
			case 3:
				int input = asJsonObject.get("inputType").getAsInt();
				String name = asJsonObject.get("name").getAsString();
				InputType inputType = new InputType(input, name);
				inputType.setValue(asJsonObject.get("value").getAsString());
				resultParts.add(inputType);
				break;
			default:
				break;
			}
		}

		return result;
	}
}
